sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -278,27 +280,159 @@ class RocksDB(
+    var changelogReader: StateStoreChangelogReader = null
+    var currLineage: Array[LineageItem] = Array.empty
+    try {
+      changelogReader = fileManager.getChangelogReader(version, stateStoreCkptId)
+      currLineage = changelogReader.lineage
+      logInfo(log"Loading lineage: " +
+        log"${MDC(LogKeys.LINEAGE, lineageManager)} from " +
+        log"changelog version: ${MDC(LogKeys.VERSION_NUM, version)} " +
+        log"uniqueId: ${MDC(LogKeys.UUID, stateStoreCkptId.getOrElse(""))}.")
+    } finally {
+      if (changelogReader != null) {
+        changelogReader.closeIfNeeded()
+      }
+    }
+    currLineage
+  }
+
+
   /**
    * Load the given version of data in a native RocksDB instance.
    * Note that this will copy all the necessary file from DFS to local disk as needed,
    * and possibly restart the native RocksDB instance.
    */
-  def load(
+  private def loadV2(
+        version: Long,
+        stateStoreCkptId: Option[String],
+        readOnly: Boolean = false): RocksDB = {
+  // An array contains lineage information from [snapShotVersion, version] (inclusive both end)
+  var currVersionLineage: Array[LineageItem] = lineageManager.getLineage()
+  try {
+    if (loadedVersion != version || (loadedStateStoreCkptId.isEmpty ||
+        stateStoreCkptId.get != loadedStateStoreCkptId.get)) {
+      closeDB(ignoreException = false)
+
+      val (latestSnapshotVersion, latestSnapshotUniqueId) = {
+        // Special handling when version is 0.
+        // When loading the very first version (0), stateStoreCkptId does not need to be defined
+        // because there won't be 0.changelog / 0.zip file created in RocksDB under v2.
+        if (version == 0) {
+          assert(stateStoreCkptId.isEmpty, "stateStoreCkptId should be empty when version is zero")
+          (0L, None)
+        // When there is a snapshot file, it is the ground truth, we can skip
+        // reconstructing the lineage from changelog file.
+        } else if (fileManager.existsSnapshotFile(version, stateStoreCkptId)) {
+          currVersionLineage = Array(LineageItem(version, stateStoreCkptId.get))
+          lineageManager.resetLineage(currVersionLineage)
+          (version, stateStoreCkptId)
+        } else {
+          currVersionLineage = getLineageFromChangelogFile(version, stateStoreCkptId) :+
+            LineageItem(version, stateStoreCkptId.get)
+          currVersionLineage = currVersionLineage.sortBy(_.version)
+          lineageManager.resetLineage(currVersionLineage)
+
+          val latestSnapshotVersionsAndUniqueId =
+            fileManager.getLatestSnapshotVersionAndUniqueIdFromLineage(currVersionLineage)
+          latestSnapshotVersionsAndUniqueId match {
+            case Some(pair) => pair
+            case None =>
+              logWarning("Cannot find latest snapshot based on lineage: "
+                + printLineageItems(currVersionLineage))
+              (0L, None)
```

Comment:
@siying 
After the refactor requested, the error "cannot_load_base_snapshot" is removed
```
def cannotFindBaseSnapshotCheckpoint(lineage: String): Throwable = {
    new SparkException (
      errorClass =
        "CANNOT_LOAD_STATE_STORE.CANNOT_FIND_BASE_SNAPSHOT_CHECKPOINT",
      messageParameters = Map("lineage" -> lineage),
      cause = null)
  }
```

When there is no valid value returned from `getLatestSnapshotVersionAndUniqueIdFromLineage`. We use (0L, None). This is for the following scenario:

1. load(0, None), load(1, id1), load(2, id2), load(3, id3)
2. commit -> now only 4_id4.zip file presents
3. maintenance -> 4_id4.zip file presents uploaded an becomes the groud truth
4. load(1, id1) -> Now `getLatestSnapshotVersionAndUniqueIdFromLineage` returns nothing because the lineage loaded from 1_id1.changelog only contains <1, id1>, but file listing returns <4, id4>
Now you have to reload from the beginning and use 0 as latestSnapshotVersion

Reply:
The handling looks OK but can you explain why `cannot_load_base_snapshot` is removed? We still need to throw it if none of 1-4 is applicable?

Reply:
Yea I think this is a flaw more deeply to the rocksDB changelog... Because we don't have an empty 0_id0.zip file

Now when there are no matches there could be two cases (even without file listing I believe this still persists)
1. When you load(v, id) but there is no snapshot created yet.
2. When there are snapshot created but the lineage doesn't match.

For 1. we need to just set latestSnapshot to 0 and latestSnapshotUniqueId to None for now. For 2 we should throw error. 

How about fixing it like this:
```
latestSnapshotVersionsAndUniqueId match {
            case Some(pair) => pair
            case None if currVersionLineage.head.version == 1 && !fileManager.existsSnapshotFile(1, id1)=>
              logWarning("Cannot find latest snapshot based on lineage: "
                + printLineageItems(currVersionLineage))
              (0L, None)
            case None => throw cannot_load_base_snapshot
          }
```
When the first version is 1 in `currentLineage`, it means there is no snapshot created yet, and use 0 as default. In other cases we throw.

This looks a bit messy tho.

Reply:
@WweiL whether throwing error or not is determined by whether we are able to load a correct checkpoint. The condition should be simple:
1. we have valid snapshot for version v, and all valid changelog fils v+1, v+2, ..., versionToLoad, or
2. we have changelog file 1, 2, 3, ... versionToLoad.
If we cannot satisfy it, there is no way we can load a checkpoint, so we should throw an error.

Reply:
I think your proposed change is good if you remove the `&& !fileManager.existsSnapshotFile(1, id1)` part.

Reply:
Thanks, changed